### PR TITLE
[RNG] Add rocRAND backend

### PR DESCRIPTION
**Pull request by sbalint98**
# Description

This PR adds support for the rocRAND backend via hipSYCL. This PR is independent of #144 and can be merged independently.

Note: Unless #149 is resolved in order to test this PR with hipSYCL the [half_patch.txt](https://github.com/oneapi-src/oneMKL/files/7891872/half_patch.txt) needs to be applied that switches to the `cl::sycl::` namespace.

Note 2: Due to a bug in rocRAND in order to pass all the tests, one needs to manually apply https://github.com/ROCmSoftwarePlatform/rocRAND/pull/239 to the rocRAND develop and compile rocRAND manually. Otherwise the `skip_ahead` tests will fail.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
- [x] Have you formatted the code using clang-format?

[rng-hipsycl-curand-test.txt](https://github.com/oneapi-src/oneMKL/files/7891868/rng-hipsycl-curand-test.txt)
[rng-dpcpp-cpu-curand-test.txt](https://github.com/oneapi-src/oneMKL/files/7891869/rng-dpcpp-cpu-curand-test.txt)
[rng-hipsycl-rocrand-test.txt](https://github.com/oneapi-src/oneMKL/files/7891870/rng-hipsycl-rocrand-test.txt)

### CODE REVIEW

File: include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp
Comment (aelizaro) [inline include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp]: Should it be ROCrand backend in License?

File: include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp
Comment (aelizaro) [inline include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp]: CURAND->ROCRAND in guard?

File: src/include/function_table_initializer.hpp
Comment (aelizaro) [inline src/include/function_table_initializer.hpp]: I think `cout` here was added for the debug purpose

File: src/rng/backends/common/hiplike_helper.hpp
Comment (aelizaro) [inline src/rng/backends/common/hiplike_helper.hpp]: curand->rocrand in this file, or should it be common for both curand and rocrand, then probably it makes sense to put it into common/ includes on src/rng level

File: tests/unit_tests/main_test.cpp
Comment (aelizaro) [inline tests/unit_tests/main_test.cpp]: extra `cout`?

File: src/rng/backends/common/hiplike_helper.hpp
Comment (sbalint98) [reply src/rng/backends/common/hiplike_helper.hpp]: Yes, this file is common for both the rocRAND and the cuRAND backend. I wonder, wouldn't it make more sense to put into /include/oneapi/mkl/rng/detail?

Comment (aelizaro) [reply src/rng/backends/common/hiplike_helper.hpp]: Yes, I think it is also fine to do, as it may be used for any other implementations, which doesn't have such scaling

**Comment by sbalint98** [comment]
Thanks, for the comments @aelizaro I hope I have addressed everything, as expected the tests are still passing:
[rng-hipsycl-rcorand-test2.txt](https://github.com/oneapi-src/oneMKL/files/7908594/rng-hipsycl-rcorand-test2.txt)
[rng-hipsycl-curand-test2.txt](https://github.com/oneapi-src/oneMKL/files/7908595/rng-hipsycl-curand-test2.txt)
[rng-dpcpp-curand-test2.txt](https://github.com/oneapi-src/oneMKL/files/7908596/rng-dpcpp-curand-test2.txt)

**Review by aelizaro — APPROVED** [review]
Looks good to me!

File: tests/unit_tests/main_test.cpp
Comment (mkrainiuk) [inline tests/unit_tests/main_test.cpp]: Looks like this new line is not needed, could you please remove it?

**Review by mkrainiuk — APPROVED** [review]
Thank for the PR! Changes look good to me, just one minor comment.

**Comment by mmeterel** [comment]
@sbalint98 I think README needs to be updated for this PR as well. 
@aelizaro Could you please test the PR locally on the future CI machine?

**Comment by aelizaro** [comment]
@mmeterel, I have tested PR locally, everything is ok (I have only added sycl::half w/a which is part of https://github.com/oneapi-src/oneMKL/pull/144 )
[RocRAND_test.txt](https://github.com/oneapi-src/oneMKL/files/8042572/RocRAND_test.txt)

**Comment by mmeterel** [comment]
> @mmeterel, I have tested PR locally, everything is ok (I have only added sycl::half w/a which is part of #144 ) [RocRAND_test.txt](https://github.com/oneapi-src/oneMKL/files/8042572/RocRAND_test.txt)

@aelizaro thank you. We are seeing a build issue with hipSYCL after rebase. After that issue is resolved, #144 will be good to merge (for me :))

File: include/oneapi/mkl/detail/get_device_id.hpp
Comment (mmeterel) [inline include/oneapi/mkl/detail/get_device_id.hpp:29]: Looks duplicate.

File: include/oneapi/mkl/rng/detail/hiplike_helper.hpp
Comment (mmeterel) [inline include/oneapi/mkl/rng/detail/hiplike_helper.hpp:2]: I am not sure adding rocRAND into this Copyright makes sense. I understand this is a common file between rocRAND and cuRAND, but the contributions are done by different developers. I would think they will need separate copyrights.

File: include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp
Comment (mmeterel) [inline include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp:2]: Confused here. Why rocRAND is using this copyright?

File: src/rng/backends/curand/curand_common.hpp
Comment (mmeterel) [inline src/rng/backends/curand/curand_common.hpp:2]: Why does this have this copyright?

File: src/rng/backends/rocrand/CMakeLists.txt
Comment (mmeterel) [inline src/rng/backends/rocrand/CMakeLists.txt:2]: why rocRAND has this copyright?

File: src/rng/backends/rocrand/mkl_rng_rocrand_wrappers.cpp
Comment (mmeterel) [inline src/rng/backends/rocrand/mkl_rng_rocrand_wrappers.cpp:2]: same here

File: src/rng/backends/rocrand/rocrand_common.hpp
Comment (mmeterel) [inline src/rng/backends/rocrand/rocrand_common.hpp:2]: Why this copyright?

File: src/rng/backends/rocrand/rocrand_helper.hpp
Comment (mmeterel) [inline src/rng/backends/rocrand/rocrand_helper.hpp:2]: same here

**Review by mmeterel — COMMENTED** [review]
Thanks for the PR. A few notes:

- Please check if the copyrights are added properly. 
- Please update README.

**Comment by illuhad** [comment]
@mmeterel We are unsure how to proceed regarding the copyright. we have based the rocRAND backend on the existing cuRAND backend, which has this license. As such, it is a derived work. The license of the cuRAND backend explicitly states that the copyright notice must not be removed. I guess nobody here is a lawyer, but we wonder if Intel legal team has experience with how to handle such issues?

Overall, I think that contributing to oneMKL might be easier in that regard if there were a common license and/or contributor CLA.

**Comment by cdgarland** [comment]
> @mmeterel We are unsure how to proceed regarding the copyright. we have based the rocRAND backend on the existing cuRAND backend, which has this license. As such, it is a derived work. The license of the cuRAND backend explicitly states that the copyright notice must not be removed. I guess nobody here is a lawyer, but we wonder if Intel legal team has experience with how to handle such issues?
> 
> Overall, I think that contributing to oneMKL might be easier in that regard if there were a common license and/or contributor CLA.

@TejaX-Alaghari This project is licensed und the Apache License Version 2.0. Any contribution of code into this project must be compatible with and able to be redistributed under that license. We will need to have to have a review of the license/copyright that you are submitted to verify this before we can merge it in.

**Comment by illuhad** [comment]
@cdgarland Thanks. To clarify, we are not adding any new license - the license in question is the one of the cuRAND backend, which we have based our changes on and which already exists in oneMKL.

**Comment by mmeterel** [comment]
> @cdgarland Thanks. To clarify, we are not adding any new license - the license in question is the one of the cuRAND backend, which we have based our changes on and which already exists in oneMKL.

@illuhad Sorry for the delay. After some internal review, we would recommend the following copyright for the files that had LBNL copyright originally. Please let us know if you have any concerns. 

/*******************************************************************************
* Copyright (C) 2022 Heidelberg University, Engineering Mathematics and Computing Lab (EMCL)
* and Computing Centre (URZ)
* cuRAND back-end Copyright (c) 2021, The Regents of the University of
* California, through Lawrence Berkeley National Laboratory (subject to receipt
* of any required approvals from the U.S. Dept. of Energy). All rights
* reserved.
*
.......Rest of the existing copyright

File: include/oneapi/mkl/rng/detail/hiplike_helper.hpp
Comment (mmeterel) [reply include/oneapi/mkl/rng/detail/hiplike_helper.hpp:2]: Please update as suggested.

File: include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp
Comment (mmeterel) [reply include/oneapi/mkl/rng/detail/rocrand/onemkl_rng_rocrand.hpp:2]: Please update as suggested.

File: src/rng/backends/rocrand/CMakeLists.txt
Comment (mmeterel) [reply src/rng/backends/rocrand/CMakeLists.txt:2]: Please update as suggested.

File: src/rng/backends/rocrand/mkl_rng_rocrand_wrappers.cpp
Comment (mmeterel) [reply src/rng/backends/rocrand/mkl_rng_rocrand_wrappers.cpp:2]: Please update as suggested.

File: src/rng/backends/rocrand/rocrand_helper.hpp
Comment (mmeterel) [reply src/rng/backends/rocrand/rocrand_helper.hpp:2]: Please update as suggested.

File: src/rng/backends/curand/curand_common.hpp
Comment (mmeterel) [reply src/rng/backends/curand/curand_common.hpp:2]: This looks like a new file. You can use your own copyright I believe.

File: src/rng/backends/rocrand/rocrand_common.hpp
Comment (mmeterel) [reply src/rng/backends/rocrand/rocrand_common.hpp:2]: This looks like a new file. You can use your own copyright I believe.

**Comment by mmeterel** [comment]
@sbalint98 
I added more comments regarding copyright notices. There are also other small update requests. Could you please help with them?

**Comment by sbalint98** [comment]
@mmeterel Sorry for the long silence I am working on it, the rocRAND backend works fine, but I will have to make a few changes on the CPU and CUDA ones.

**Pull request closed without merging**
